README.md
# PRN-maps
Map visualisations for PRN data

# Getting started
To run the app locally, checkout the repo and run:
1. `npm install`
0. `npm start` to run an express server on http://localhost:3735

# Publishing
The site is published with GitHub pages at https://prn-maps.planetaryresponsenetwork.org/
Updating the master branch should publish changes automatically.

# Browser support
The code uses ES6 syntax, which should work in most up-to-date browsers but may not work in Internet Explorer.


CHANGELOG.md
# Change Log

## [Unreleased](https://github.com/zooniverse/PRN-maps/tree/HEAD)

**Closed issues:**

- Allow users to toggle multiple layers at once [\#8](https://github.com/zooniverse/PRN-maps/issues/8)
- Add PRN events overview landing page [\#6](https://github.com/zooniverse/PRN-maps/issues/6)

**Merged pull requests:**

- Filter layer data by minimum value [\#13](https://github.com/zooniverse/PRN-maps/pull/13) ([eatyourgreens](https://github.com/eatyourgreens))
- Multiple heatmap layers [\#11](https://github.com/zooniverse/PRN-maps/pull/11) ([eatyourgreens](https://github.com/eatyourgreens))
- Build events list from the API [\#10](https://github.com/zooniverse/PRN-maps/pull/10) ([eatyourgreens](https://github.com/eatyourgreens))
- Add events listing [\#9](https://github.com/zooniverse/PRN-maps/pull/9) ([eatyourgreens](https://github.com/eatyourgreens))
- Map options [\#4](https://github.com/zooniverse/PRN-maps/pull/4) ([eatyourgreens](https://github.com/eatyourgreens))
- Add a note about browser support [\#3](https://github.com/zooniverse/PRN-maps/pull/3) ([eatyourgreens](https://github.com/eatyourgreens))
- add getting started notes [\#2](https://github.com/zooniverse/PRN-maps/pull/2) ([camallen](https://github.com/camallen))
- Connect up the events API [\#1](https://github.com/zooniverse/PRN-maps/pull/1) ([eatyourgreens](https://github.com/eatyourgreens))



\* *This Change Log was automatically generated by [github_changelog_generator](https://github.com/skywinder/Github-Changelog-Generator)*

js/api.js
const API = {
  host: 'https://maps-api.planetaryresponsenetwork.org',
  get: function(path, defaultValue) {
    return superagent.get(`${API.host}${path}`)
    .withCredentials()
    .then(function (response) {
      return JSON.parse(response.text);
    })
    .catch(function (error) {
      console.error(error);
      return defaultValue;
    });
  },
  post: function(path, data) {
    return superagent.post(`${API.host}${path}`)
    .send(data)
    .withCredentials()
    .then(function (response) {
      return JSON.parse(response.text);
    })
    .catch(function (error) {
      console.error(error);
      return defaultValue;
    });
  },
  approve: function (eventName) {
    return API.post(`/pending/layers/${eventName}/approve`);
  },
  events: function () {
    return API.get('/events', []);
  },
  layers: function (eventName) {
    return API.get(`/layers/${eventName}`, []);
  },
  manifests: function (eventName) {
    return API.get(`/manifests/${eventName}`, {});
  },
  pendingLayers: function (eventName) {
    return API.get(`/pending/layers/${eventName}`, []);
  }
}

js/maps.js
const MAP_CONTAINER = document.getElementById('map');
const MAP_THRESHOLD = document.getElementById('map-threshold');
const MAP_SELECT = document.getElementById('map-select');
const MAP_OPTIONS = {
  zoom: 10,
  mapTypeControl: true,
  mapTypeControlOptions: {
    style: google.maps.MapTypeControlStyle.DROPDOWN_MENU,
    mapTypeIds: ['hybrid', 'satellite', 'roadmap', 'terrain']
  },
  mapTypeId: 'hybrid',
  scaleControl: true,
  streetViewControl: false
}
const HEATMAPS = {};
const HEATMAP_DATA = {};

let HEATMAP_COLOURS = [  // Pre-set colour gradients for each heatmap
  ['rgba(204, 068, 068, 0.0)', 'rgba(204, 068, 068, 0.6)', 'rgba(255, 136, 136, 0.6)'],  // Red
  ['rgba(204, 204, 068, 0.0)', 'rgba(204, 204, 068, 0.6)', 'rgba(255, 255, 136, 0.6)'],  // Yellow
  ['rgba(068, 204, 204, 0.0)', 'rgba(068, 204, 204, 0.6)', 'rgba(136, 255, 255, 0.6)'],  // Cyan
  ['rgba(068, 204, 068, 0.0)', 'rgba(068, 204, 068, 0.6)', 'rgba(136, 255, 136, 0.6)'],  // Green
  ['rgba(204, 068, 204, 0.0)', 'rgba(204, 068, 204, 0.6)', 'rgba(255, 136, 255, 0.6)'],  // Magenta
  ['rgba(068, 068, 204, 0.0)', 'rgba(068, 068, 204, 0.6)', 'rgba(136, 136, 255, 0.6)'],  // Blue
];
// Pad the colour gradient to favour the high-intensity colours.
// We need at least 8 steps in the gradient - the first being transparent - for this to look good.
HEATMAP_COLOURS = HEATMAP_COLOURS.map(function spreadColours(arr) {
  return [arr[0], arr[1], arr[1], arr[1], arr[2], arr[2], arr[2], arr[2]];
});


function queryParams() {
  const queryString = window.location.search.substring(1);
  const queryPairs = queryString.split('&');
  return queryPairs.reduce(function (query, queryPair) {
    const [param, value] = queryPair.split('=');
    query[param] = value;
    return query;
  }, {});
}

function buildLayerInput(layer) {
  const option = document.createElement('label');
  const checkbox = document.createElement('input');
  const text = document.createTextNode(layer.name)
  checkbox.type='checkbox';
  checkbox.value = layer.url;
  checkbox.checked = true;
  option.appendChild(checkbox)
  option.appendChild(text);
  return option;
}
function buildLayersMenu(layers) {
  document.querySelectorAll('#map-select label').forEach(function (node) {
    MAP_SELECT.removeChild(node);
  });
  layers
    .map(buildLayerInput)
    .forEach(function (input) {
      MAP_SELECT.appendChild(input);
    });
}

const eventName = queryParams().event
const getLayers = queryParams().pending ? 'pendingLayers' : 'layers';
API[getLayers](eventName)
  .then(buildLayersMenu)
  .then(renderMapAndFit);

const center = new google.maps.LatLng(15.231458142,-61.2507115);
const map = new google.maps.Map(MAP_CONTAINER, Object.assign(MAP_OPTIONS, { center }));

let bounds;
let fitToBounds = false;

function minimumWeight([lat, lng, weight]) {
  const threshold = parseInt(MAP_THRESHOLD.value);
  return weight > threshold;
}

function parseLine([lat, lng, weight]) {
  const location = new google.maps.LatLng(lat, lng);
  bounds.extend(location);
  return { location, weight };
}

function filteredMapData(results) {
  return results.data
    .filter(minimumWeight)
    .map(parseLine);
}

function parseMapData(results, url) {
  // For each heatmap, assign a preset colour to it.
  const numberOfHeatmaps = Object.keys(HEATMAPS).length;
  const colourIndex = numberOfHeatmaps % HEATMAP_COLOURS.length;
  const heatmapColour = HEATMAP_COLOURS[colourIndex];
  
  const heatmap = new google.maps.visualization.HeatmapLayer({
    gradient: heatmapColour,
    maxIntensity: 30,
    opacity: 1
  });
  bounds  = new google.maps.LatLngBounds();
  const heatmapData = filteredMapData(results);
  heatmap.setData(heatmapData);
  if (fitToBounds) {
    map.fitBounds(bounds);
    map.panToBounds(bounds);
    fitToBounds = false;
  }
  heatmap.setMap(map);
  HEATMAPS[url] = url ? heatmap : undefined;
}

function cacheMapData(results, file) {
  const url = file.streamer._input;
  console.log(url)
  HEATMAP_DATA[url] = url ? results : undefined;
  parseMapData(results, url); 
}

function readMapFile(url) {
  const config = {
    download: true,
    fastMode: true,
    skipEmptyLines: true,
    chunk: cacheMapData
  }
  console.log(url)
  Papa.parse(url, config);
}

function renderMap() {
  const urls = Object.keys(HEATMAPS);
  urls.forEach(function (url) {
    HEATMAPS[url] && HEATMAPS[url].setMap(null);
  })
  MAP_SELECT.querySelectorAll('input:checked')
    .forEach(function (node) {
      const url = node.value;
      if (HEATMAPS[url]) {
        const heatmapData = filteredMapData(HEATMAP_DATA[url]);
        HEATMAPS[url].setData(heatmapData);
        HEATMAPS[url].setMap(map);
      } else {
        readMapFile(url);
      }
    })
}

function renderMapAndFit() {
  fitToBounds = true;
  renderMap();
}

MAP_SELECT.addEventListener('change', renderMapAndFit);
MAP_THRESHOLD.addEventListener('change', renderMap);


js/app.js
const MAP_CONTAINER = document.getElementById('map');
const MAP_THRESHOLD = document.getElementById('map-threshold');
const MAP_SELECT = document.getElementById('map-select');
const MAP_OPTIONS = {
  zoom: 10,
  mapTypeControl: true,
  mapTypeControlOptions: {
    style: google.maps.MapTypeControlStyle.DROPDOWN_MENU,
    mapTypeIds: ['hybrid', 'satellite', 'roadmap', 'terrain']
  },
  mapTypeId: 'hybrid',
  scaleControl: true,
  streetViewControl: false
}
const HEATMAPS = {};
var TILE_URL = 'http://local.zooniverse.org:8080/data/palu_2018/{z}/{x}/{y}.png';

const layerID = 'palu_2018_before';

function queryParams() {
  const queryString = window.location.search.substring(1);
  const queryPairs = queryString.split('&');
  return queryPairs.reduce(function (query, queryPair) {
    const [param, value] = queryPair.split('=');
    query[param] = value;
    return query;
  }, {});
}

function buildLayersMenu(layers) {
  document.querySelectorAll('#map-select option').forEach(function (node) {
    MAP_SELECT.removeChild(node);
  });
  layers.forEach(function (layer, i) {
    const option = document.createElement('option');
    option.value = layer.url;
    option.text = layer.name;
    MAP_SELECT.add(option);
  });
}

const eventName = queryParams().event
const getLayers = queryParams().pending ? 'pendingLayers' : 'layers';
API[getLayers](eventName)
.then(buildLayersMenu)
.then(renderMapAndFit);

const center = new google.maps.LatLng(15.231458142,-61.2507115);
const map = new google.maps.Map(MAP_CONTAINER, Object.assign(MAP_OPTIONS, { center }));

// create the custom map tiles
var customTileMap = new google.maps.ImageMapType({
  name: layerID,
  getTileUrl: function(coord, zoom) {
    // console.log(coord);
    var url = TILE_URL
      .replace('{x}', coord.x)
      .replace('{y}', coord.y)
      .replace('{z}', zoom);
    return url;
  },
  tileSize: new google.maps.Size(256, 256),
  minZoom: 10,
  maxZoom: 15
});

// add the tile map type as an overlay over the existing
map.overlayMapTypes.insertAt(0, customTileMap);

const heatmap = new google.maps.visualization.HeatmapLayer({
  map,
  maxIntensity: 30,
  opacity: .4
});
let bounds;
let heatmapData;
let fitToBounds = false;

function minimumWeight([lat, lng, weight]) {
  const threshold = parseInt(MAP_THRESHOLD.value);
  return weight > threshold;
}

function parseLine([lat, lng, weight]) {
  const location = new google.maps.LatLng(lat, lng);
  bounds.extend(location);
  return { location, weight };
}

function parseMapData(results) {
  bounds  = new google.maps.LatLngBounds();
  heatmapData = results.data
    .filter(minimumWeight)
    .map(parseLine);
  heatmap.setData(heatmapData);
  if (fitToBounds) {
    map.fitBounds(bounds);
    map.panToBounds(bounds);
    fitToBounds = false;
  }
}

function cacheMapData(results, file) {
  const url = MAP_SELECT.value;
  HEATMAPS[url] = url ? results : undefined;
  parseMapData(results);
}

function readMapFile(url) {
  const config = {
    download: true,
    fastMode: true,
    skipEmptyLines: true,
    chunk: cacheMapData
  }
  console.log(url)
  Papa.parse(url, config);
}

function renderMap() {
  const url = MAP_SELECT.value;
  if (HEATMAPS[url]) {
    parseMapData(HEATMAPS[url]);
  } else {
    readMapFile(url);
  }
}

function renderMapAndFit() {
  fitToBounds = true;
  renderMap();
}

MAP_SELECT.addEventListener('change', renderMapAndFit);
MAP_THRESHOLD.addEventListener('change', renderMap);
